### ISSUE TITLE
'require' doesn't work for npm modules from parent folder

### ISSUE THREAD

**Issue by bisubus** [ivogabe/small#6]
I've ran into the issue with npm 2.x and hierarchical `node_modules` [here](https://github.com/ex-machine/ng-debottle/tree/1.1.0)

The package has `apply-fn` and `debottle` dependencies, the latter depends on `apply-fn` too, so `node_modules/debottle/node_modules/apply-fn` isn't created during package installation, and Node expects that `node_modules/apply-fn` will be used for `requre('apply-fn')` in all modules.

When the package is bundled with Small this results in

```
Reference not found:
        "apply-fn" from <current_dir>\node_modules\debottle\src\debottle.js
```

It looks like Small doesn't look up for the module in parent folders, like Node would in this case.

Building is handled by Gulp like that:

```
gulp.task('build', () => {
    return gulp.src(['src/**.js'])
    .pipe(gulp.$.small(packageName + '.js', {
        externalResolve: ['node_modules'],
        exportPackage: {
            universal: packageName 
        },
        outputFileName: {
            universal: packageName + '.js'
        },
    }))
    .pipe(gulp.dest('dist'));
});
```

Small 0.1.1, node 5.1.1, npm 2.14.8 on Windows.

And Small 0.2.x just keeps throwing

```
Cannot read property 'kind' of undefined
```

with no call stack.

**Comment by ivogabe** [comment]
I think that error has been fixed in 0.2. I'm not sure why you're getting that error. I'll investigate it with your repository.

**Comment by bisubus** [comment]
It looks like that 0.2.2 throws the exception on the sources that have any JS code (but empty file and the one that only consists of `'use strict'; (function () {})();` are ok). I found that call stack is being outputted for `small` CLI, hope this helps.

```
Unhandled rejection TypeError: Cannot read property 'kind' of undefined
    at Object.isTypeNode (...\node_modules\small\node_modules\typescript\lib\typescript.js:5526:32)
    at Object.getTypeOfNode [as getTypeAtLocation] (...\node_modules\small\node_modules\typescript\lib\typescript.js:26314:20)
    at Function.ExportNode.tryParse (...\node_modules\small\release\node.js:61:124)
    at Function.ExportNode.tryParse (...\node_modules\small\release\node.js:93:47)
    at Walker.callback (...\node_modules\small\release\parser.js:37:48)
    at Walker.visit (...\node_modules\small\release\parser.js:113:14)
    at ...\node_modules\small\release\parser.js:121:63
    at visitNode (...\node_modules\small\node_modules\typescript\lib\typescript.js:7405:20)
    at Object.forEachChild (...\node_modules\small\node_modules\typescript\lib\typescript.js:7583:24)
    at Walker.descend (...\node_modules\small\release\parser.js:121:12)
    at Walker.callback (...\node_modules\small\release\parser.js:44:20)
    at Walker.visit (...\node_modules\small\release\parser.js:113:14)
    at ...\node_modules\small\release\parser.js:121:63
    at visitEachNode (...\node_modules\small\node_modules\typescript\lib\typescript.js:7417:30)
    at Object.forEachChild (...\node_modules\small\node_modules\typescript\lib\typescript.js:7574:24)
    at Walker.descend (...\node_modules\small\release\parser.js:121:12)
```

**Comment by ivogabe** [comment]
I've cloned your repository and ran `npm install`, `npm install small@0.2.2` and `gulp`. I got no errors though. Can you try to remove `node_modules` and install the dependencies again?

**Comment by bisubus** [comment]
I was able to replicate the issue (Cannot read property 'kind' of undefined) with fresh package

```
.\node_modules\.bin\small -i test.js -o output.js
```

and test.js is

```
var test;
```

Have you checked how it performs on Windows by chance? The behaviour is consistent on Windows 7, with arbitrary Node 4.x and 5.x versions

I've also checked 0.2.0, and the exception makes more sense there

```
Possibly unhandled TypeError: Cannot read property 'indexOf' of undefined
    at hasExtension (...\node_modules\small\node_modules\typescript\lib\typescript.js:36540:48)
    at processSourceFile (...\node_modules\small\node_modules\typescript\lib\typescript.js:36598:17)
    at processRootFile (...\node_modules\small\node_modules\typescript\lib\typescript.js:36543:13)
    at Object.createProgram (...\node_modules\small\node_modules\typescript\lib\typescript.js:36291:17)
    at Parser.parse (...\node_modules\small\release\parser.js:24:26)
    at ...\node_modules\small\release\project.js:69:26
    at tryCatch1 (...\node_modules\small\node_modules\bluebird\js\main\util.js:63:19)
    at Promise$_callHandler [as _callHandler] (...\node_modules\small\node_modules\bluebird\js\main\promise.js:695:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (...\node_modules\small\node_modules\bluebird\js\main\promise.js:711:18)
    at Promise$_settlePromiseAt [as _settlePromiseAt] (...\node_modules\small\node_modules\bluebird\js\main\promise.js:868:14)
```

It was clearly caused by non-normalized path in `hasExtension`, etc. And it works fine when the path is normalized.

**Comment by ivogabe** [comment]
In the latest version, 0.2.5, I fixed a bug on Windows. #8 was probably the same issue that you had, could you check that?

**Comment by bisubus** [comment]
Thanks, it works for Windows now.

The difference I've noticed:

```
var apply = require('apply-fn');
apply();
```

in 0.1.1 is built to 

```
...
var exports = {};
var apply = ((function() {
var exports = {};
function apply(self, fn, args) { ... }

exports = apply;
return exports;
})());

apply();
...
```

and in 0.2.5 is built to 

```
...
var exports = {};
var __small$_1 = (function() {
var exports = {};
function apply(self, fn, args) { ... }

exports = apply;
return exports;
})();
var __small$_1;

apply();
...
```

0.2.5 has `var __small$_1` variable redeclaration. And `apply` is not defined, so it will throw. Should it be moved to a new issue?

**Comment by ivogabe** [comment]
The change to `__small$_1` was intended, but `apply()` should be converted to `__small$_1()`. That works correctly on my Mac, but on Windows it apparently not... 😞
I'll investigate this.

**Comment by ivogabe** [comment]
I think I've fixed it with cf89fba6a8b2543500962ae626141ce5141b5302.

**Comment by bisubus** [comment]
Yes, I've checked it out, everything looks great so far, thanks.
